graf2d/gpadv7/inc/ROOT/RPadBase.hxx

```diff
@@ -81,43 +83,45 @@ public:
    /// \returns vector of vector (ret[x][y]) of created pads.
    std::vector<std::vector<std::shared_ptr<RPad>>> Divide(int nHoriz, int nVert, const RPadExtent &padding = {});
 
-   /// Create drawable of specified class T
-   template<class T, class... ARGS>
-   auto Draw(ARGS... args)
+   /// Add object to be painted.
+   /// Correspondent drawable will be created via GetDrawable() function which should be defined and be accessed at calling time.
+   /// If required, extra arguments for GetDrawable() function can be provided.
+   template <class T, class... ARGS>
+   auto Draw(const std::shared_ptr<T> &what, ARGS... args)
    {
-      auto drawable = std::make_shared<T>(args...);
+      // Requires GetDrawable(what) to be known!
+      auto drawable = GetDrawable(what, args...);
 
       TestIfFrameRequired(drawable.get());
 
-      fPrimitives.emplace_back(drawable);
+      AddPrimitive(drawable);
```

Comment:
Would it make sense to pass the owning pad into `GetDrawable()`, rather than depending on `AddPrimitive()` to do something special for pads and only for pads? That way, `GetDrawable<RPad>` can do "the right thing".

Reply:
Yes, one can hide `RPad` constructor and provide specialization for `RPadBase::Draw<RPad>()`.
I prefer to use this signature and not `GetDrawable()` function.

Reply:
The advantage of using `GetDrawable` is that this is *meant* to be customized by different classes. Here you need to inject a customization ("hack") for `RPad` to interject and set the parent. This is surprising: if I were to ask a future team member, where do you think `SetParent` is called - they would not know!

`friend` declarations can often be a symptom of the violation of separation of concerns.

In general I could imaging that it's useful for `GetDrawable()` to know the pad it will be painted on. Not just for `RPad`'s purposes.

Reply:
I could do it differently - just hide `RPad` constructor and let create it only via `RPadBase::CreateSubPad()` method.
Like we do it with `RFrame`.

Reply:
I cannot tell what this change would mean for users? I suppose we would prevent them from calling `Draw<RPad>(...)`? I think that's actually nice.

Can I have some counter-arguments against what I'm proposing, please? :-) I understand you prefer your approach - what's wrong with changing the `GetDrawable()` signature?

Reply:
> what's wrong with changing the GetDrawable() signature

Everywhere extra `RPadBase &pad` argument which not used in 99% of cases.

Reply:
Thanks. I wouldn't extrapolate from what we have now to what we'll have in 5 years. I can imagine more complex objects to adjust their representation based on the aspect ratio of the pad, or they might automatically create sub-pads inside the pad (think `TRatioPlot`) etc.

We *could* check whether the overload `GetDrawable(RPad&, shared_ptr<T>&,...)` exists, else call `GetDrawable(shared_ptr<T>&,...)`. What do you think?

Reply:
> We could check whether the overload `GetDrawable(RPad&, shared_ptr<T>&,...)` exists, 
> else call `GetDrawable(shared_ptr<T>&,...)`. What do you think?

Would be ok, but we always should remember about `python` binding - PyROOT not always perfect with handling of complex C++ templates. 

Probably we also should use different names for `canv->Draw(hist)` which calls `GetDrawable()` and 
`canv->Draw<RLine>(p1,p2)`. May be use `canv->Add<RLine>(p1,p2)`?

Reply:
`GetDrawable()` isn't called from the Python side - it's called by `RPadBase::Draw()`, right?

A agree on the change to`Add`, or even `Create` (similarly to `RDirectory`)?

Reply:
> `GetDrawable()` isn't called from the Python side - it's called by `RPadBase::Draw()`, right?

Yes, but `RPadBase::Draw` is templated method in this case and I have no idea to which extend PyROOT can unfold such templates, where `GetDrawable` is of main difficulty for it.

Reply:
Python only interacts with `RPadBase::Draw()` and not with functions called by it.

Reply:
> Python only interacts with `RPadBase::Draw()` and not with functions called by it.

But even cling has difficulty to properly load function when calling `canv->Draw(hist)`.
In many v7 macros we have calls `R__LOAD_LIBRARY(libROOTHistDraw)`
How PyROOT can survive it?

Reply:
PyROOT is invoking the function through cling. If cling has an issue, fine, we need to fix this (and the missing symbols *should* be addressed by now in master and 6.24, especially with modules). But PyROOT itself doesn't see the function body of what it invokes. 

Maybe we're talking past one another, but my understanding is that Python code might invoke `canv.Draw(foo)`. That is translated into a call in cling - and the function body of the inlined `RCanvas::Draw()` is "compiled" by cling - PyROOT doesn't deal with that. *If* Python code were to call `GetDrawable()` then it would have to do SFINAE + overload selection, but that doesn't seem to be a realistic case.